src/mgr/ActivePyModules.cc

```diff
@@ -985,6 +1094,161 @@ PyObject* ActivePyModules::get_perf_schema_python(
+    daemons = daemon_state.get_all();
+  } else if (svc_id.empty()) {
+    daemons = daemon_state.get_by_service(svc_type);
+  } else {
+    auto key = DaemonKey{svc_type, svc_id};
+    // so that the below can be a loop in all cases
+    auto got = daemon_state.get(key);
+    if (got != nullptr) {
+      daemons[key] = got;
+    }
+  }
+
+  auto f = with_gil(no_gil, [&] {
+    return PyFormatter();
+  });
+
+  auto dump_sub_counter_information = [] (PyFormatter *f, PerfCounterType type){
+    Formatter::ObjectSection counter_section(*f, type.counter_name);
+    f->create_unique("description", type.description);
+    if (!type.nick.empty()) {
+      f->dump_string("nick", type.nick);
+    }
+    f->dump_unsigned("type", type.type);
+    f->dump_unsigned("priority", type.priority);
+    f->dump_unsigned("units", type.unit);
+  };
+
+  auto dump_counter_with_labels = [&dump_sub_counter_information] (PyFormatter *f, auto key_labels, auto type){
+    f->open_object_section(""); // counter should be enclosed by array
+
+    for (Formatter::ObjectSection labels_section{*f, "labels"}; const auto &label: key_labels){
+      f->dump_string(label.first, label.second);
+    }
+
+    f->open_object_section("counters");
+    dump_sub_counter_information(f, type);
+  };
+
+
+  if (!daemons.empty()) {
+    for (auto& [key, state] : daemons) {
+      std::lock_guard l(state->lock);
+      with_gil(no_gil, [&, key=ceph::to_string(key), state=state] {
+        string_view key_name, prev_key_name;
+        vector<pair<string_view,string_view>> prev_key_labels;
+        Formatter::ObjectSection counter_section(f, key.c_str()); // Main Object Section
+        std::optional<Formatter::ArraySection> array_section;
+
+        for (auto ctr_inst_iter : state->perf_counters.instances) {
+          /*
+              The path of the counter can either be:
+                - labeled counter path: "osd_scrub_sh_repl^@level^@shallow^@pooltype^@replicated^@.successful_scrubs_elapsed"
+                - unlabeled counter path: "osd.stat_bytes"
+              
+              For the above counters:
+                - key_names are: 'osd_scrub_sh_repl' and 'osd'
+                - counter names are: 'successful_scrubs_elapsed' and 'stat_bytes'
+
+          */
+          const auto &counter_name_with_labels = ctr_inst_iter.first;
+          auto type = state->perf_counters.types[counter_name_with_labels];
+
+          // create a vector of labels i.e [(level, shallow), (pooltype, replicated)]
+          vector<pair<string_view,string_view>> key_labels;
+          for (auto label : ceph::perf_counters::key_labels(counter_name_with_labels)) {
+            if (!label.first.empty()){
+              key_labels.push_back(label);
+            }
+          }
+
+          // Extract the key names from the counter path, these key names form
+          // the main object section for their counters
+          string key_name_without_counter;
+          if(!key_labels.empty()){
```

Comment:
swap the two sides of the fork, instead of using a double negative ('not' and 'empty')

Reply:
done!